### PR TITLE
Ensure test inputs are downloaded before opening the shell

### PR DESCRIPTION
**Pull request by peterebden**
Most of this is commonising other bits of code.

### CODE REVIEW

File: src/core/state.go [hunk 1]
```diff
@@ -970,6 +970,43 @@ func (state *BuildState) ForConfig(config ...string) *BuildState {
...
+		for _, datum := range target.AllData() {
+			ch <- datum
+		}
+		for _, datum := range target.TestTools() {
+			ch <- datum
+		}
+		close(ch)
+	}()
+	return ch
+}
```
Comment (Tatskaari) [inline src/core/state.go:1008]: Do we not also need to download the output of the rule i.e. the test binary?

**Review by Tatskaari — APPROVED** [review]
Looks good, just one query.

File: src/core/state.go (hunk 1 above)
Comment (peterebden) [reply src/core/state.go:1008]: I lifted this from the existing implementation in remote/action.go...

Good point but it does seem to be working fairly consistently 🤔

Comment (peterebden) [reply src/core/state.go:1008]: I think we implicitly download the test binary at the build step... 😬

**Pull request merged**